Illustrates a bug in compute=false Materials

WilkAndy opened the pull request:
This PR enhances ComputeMultipleInelasticStress by adding a 'cycle_models' flag.  When set true this flag means ComputeMultipleInelasticStress will use only inelastic model (_t_step - 1) % num_models to compute the stress.
The remainder of the compute=false models are ignored for that timestep.

Here's the bug.  The compute=false Materials set all of their internally-declared stateful MaterialProperties to zero during any timestep in which they are not called.  This is undesirable behaviour.  Instead, if not called, they should set their stateful MaterialProperties from the corresponding Old values.

This PR contains a test that will fail because of this bug.  The test computes a Postprocessor called i_cdp, derived from a stateful MaterialProperty, which is zero for the second timestep, because its corresponding compute=false Material is not called during that timestep.  The correct value is 540, which is the value of i_cdp at the end of the first timestep.

Fixes #9165

moosebuild commented:
Job `Precheck:linux-gnu` on 548605d wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 6a916d5fa569469e1c0f34189736e84854f597bf`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9170/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9170/style.patch | git apply -v`

moosebuild commented:
Job `Precheck:linux-gnu` on 270009d wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 6a916d5fa569469e1c0f34189736e84854f597bf`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9170/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9170/style.patch | git apply -v`

WilkAndy commented:
**THE DISCUSSION HERE FOLLOWS FROM THE DISCUSSION AT #9166**

WilkAndy commented:
Continuing the discussion contained in #9166...  @bwspenc wrote:

"Yeah, but what are non-stateful material properties supposed to do if they don't get computed? There is no state to copy into them. We need to be consistent with that."

OK, I understand, and hadn't thought of that case.  I'm not sure what's best:

(1) demand that the user write their own "copying" routine (that won't be a straightforward copy for non-stateful as presumably some computation will be required)

(2) make the stateful stuff automatically copied by the framework, and the user must write their own routine for non-stateful Properties.

I'm not sure what's best as both seem rather ugly!  What do the framework developers think?  @aeslaughter  , @permcody , @andrsd , @rwcarlsen  - would you care to comment?

moosebuild commented:
Job `Moose Documentation:linux-clang` on 571686b wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9170/site

rwcarlsen commented:
I don't like option 2 because it slows down code that may not want to reuse previous values at all with `compute=false`.  I think something along the lines of option 2 would probably be the best for the near term.  Rather than try to guess what users want, it is reasonable to have them do it explicitly - especially since this seems to not really be an "entry-level" type feature.

permcody commented:
I don't see how either of these options will work for non-stateful materials. If your field is spatially varying, and we don't store any of that information, how do you copy it? It seems that if you don't want to compute anything but you want the values to stay the same, you essentially need to promote (demote?) these materials to be stateful. Then perhaps the copying will be handled for you using the stateful mechanisms.

bwspenc commented:
I agree that this is not an "entry-level" feature.  In general, if you want to use a material property, I think you would expect to call some kind of function that computes it.

rwcarlsen commented:
@permcody is correct.  It doesn't make sense to reuse old values of a non-stateful material.  Regardless, I think users just need to manage these details on their own via e.g. `resetQpProperties`.

aeslaughter commented:
@rwcarlsen Is correct, this is what "resetQpProperties" is for, to make sure the properties are in a sane state for computing or use by others.

bwspenc commented:
We need to manage this on our own, but I don't think `resetQpProperties()` is the place for that to happen because it gets called all the time, not just for this very unusual special case.  I think we need to define a special method in the base class for our models of this type that copies properties from the previous state rather than doing the standard computation of them, and this would need to be explicitly called for the materials that are skipped.

WilkAndy commented:
OK, i'm disappointed that something automatic isn't being set up, but i will follow what @bwspenc suggests and create a method that derived classes can override.

WilkAndy committed:
This PR enhances ComputeMultipleInelasticStress by adding a 'cycle_models' flag.  When set true this flag means ComputeMultipleInelasticStress will use only in elastic model (_t_step - 1) % num_models to compute the stress.

The remainder of the compute=false models use only propagateQpProperties for that step.  This is a new method in StressUpdateBase that allows derived classes to propagate Old Stateful Properties forward to the current timestep (or indeed, do any other internal computations they would like to do).

Fixes #9165

moosebuild commented:
Job `Precheck:linux-gnu` on 9c4398e wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 6a916d5fa569469e1c0f34189736e84854f597bf`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9170/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9170/style.patch | git apply -v`

WilkAndy committed:
test added.  Fixes #9165

Also changes suggested by BenS

moosebuild commented:
Job `Moose Documentation:linux-clang` on a560c30 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9170/site

WilkAndy commented:
This is ready for review, @bwspenc 

a

bwspenc commented on modules/tensor_mechanics/src/materials/StressUpdateBase.C:
```diff
@@ -31,3 +31,8 @@ StressUpdateBase::setQp(unsigned int qp)
   _qp = qp;
 }
+
+void
```
I think this function should be pure virtual, and we shouldn't provide an empty implementation in the base class.

bwspenc replied:
On second thought, here's a better solution: Leave the implementation in the base class, but have it call `mooseError()`, so that if anyone tries to use the cyclic iterations with a model that doesn't support it, they'll get an error.

I think we can figure out a way to have the implementation in StressUpdateBase copy all stateful properties so we won't need to implement this in derived classes, but that will work in the mean-time.

Also, while you're making that change, how about renaming `propagateQpProperties` to `propagateQpStatefulProperties`?  That would make it clear that this only applies to stateful properties.

WilkAndy replied:
Sounds good, thanks @bwspenc ,

a

bwspenc commented on modules/tensor_mechanics/include/materials/StressUpdateBase.h:
```diff
@@ -73,6 +73,12 @@ class StressUpdateBase : public Material
+   * If updateState is not called during a timestep, this will be.  This method allows derived
+   * classes to set internal parameters from their Old values, for instance
+   */
+  virtual void propagateQpProperties();
```
I'd make this pure virtual:
`virtual void propagateQpProperties() = 0;`
That way, we will make sure that every material that derives from this will have thought about what the right thing to do here is.  There are a handful of models in TensorMechanics that will need to be updated to do this. 

We also have some models in BISON that would need to implement this.  Maybe the best way to handle this is to go ahead and merge your code as-is.  I can then add this to all of the BISON models, and then we can make it pure virtual after that is merged in.

bwspenc reviewed (approved):
I think this is all good. As I noted, I think we should make `propagateQpProperties()` pure virtual, but that will require updates to apps.

bwspenc reviewed (changes requested):
Just make the minor change noted in my comments in the previous review.

moosebuild commented:
Job `Precheck:linux-gnu` on 33bd0e6 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 6a916d5fa569469e1c0f34189736e84854f597bf`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9170/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9170/style.patch | git apply -v`

moosebuild commented:
Job `Moose Documentation:linux-clang` on ac86881 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9170/site
